### ISSUE TITLE
Inferring shapes of subsequent layers inside nn.Sequential

### ISSUE THREAD

**Issue by varenick** [pytorch/pytorch#17448]
## Feature
A modification of nn.Sequential class that would infer some input parameters for containing modules. E.g. `in_features` for `nn.Linear`.

## Motivation
Inferring shapes of subsequent layers require manual calculations. Sometimes it is not trivial; e.g. in conv layers with non-default padding / output_padding / stride. In this case one require to look into documentation page for exact formula. Of course, it is a consequence of a dynamic-graph paradigm. However in the case, when the order of layers is predefined, e.g. in `nn.Sequential` container, it seems possible and would be quite handy.

## Proposal
My idea is to first of all, implement a method `infer_output_shape` in every class inherited from `nn.Module`, which would take an input shape and return an output shape. Also, I need a method (let's call it `infer_init_params`) that would take an input shape and return a dict with arguments of `__init__`, that are inferrable from this input shape (e.g. number of input channels of `nn.Conv2d` is inferrable from input shape).

After that, I would like to extend an `nn.Sequential` class. I would like it to accept not only a sequence of `nn.Module`s, but also a sequence of either `nn.Module`s, or tuples of class and dict of kwargs, e.g. not only, say, `nn.Conv2d(3, 32, 3)`, but also `(nn.Conv2d, {in_channels=3, out_channels=32, kernel_size=3})`. The modified `nn.Sequential` class would also require `expected_input_shape` argument. The modified `nn.Sequential` class then will initialize as following: for every tuple in a sequence it will infer params to initialize a corresponding module. If all the params required to initialize a module are either inferred, or provided in a dict of kwargs, it will deduce an output shape, and infer `__init__` arguments for a next module.

A code example:
```
model = nn.Sequential(
    (nn.Linear, {out_features=500}),
    nn.ReLU(),
    (nn.Linear, {out_features=10}),
    expected_input_shape=(None,784)
)
```
Here, in `nn.Sequential.__init__` method we instantiate a `nn.Linear` object with `in_features=784` deduced from `expected_input_shape` and `out_features=500` that is given. Then we use `nn.Linear.infer_output_shape` to deduce output shape `(None, 500)`. Analogously, we use `nn.ReLU.infer_output_shape` to again deduce output shape `(None, 500)`. After that, we use `nn.Linear.infer_init_params` to deduce `{in_features=500}` from the shape `(None, 500)` and instantiate the second `nn.Linear` object with `in_features=500, out_features=10`.

Does this look too clumsy? If not, I would like to make a PR.

**Comment by soumith** [comment]
Implementing `infer_output_shape` for every class inherited from `nn.Module` is an unacceptable burden for Module writers. It's easy for `nn.Linear`, but not so easy for others.

**Comment by varenick** [comment]
@soumith It seems like for those modules, where inferring output shape is not trivial, formula for computing output shape is always given in documentation (I can be wrong, however). Hence a module-writer have to provide such a formula anyway. Given this, it shouldn't be much problem to implement `infer_output_shape`.

**Comment by soumith** [comment]
@varenick no, it's not often given in the documentation. Whatever limited form is given, doesn't actually cover all cases. Also, a lot of user-written layers simply dont have documentation as well.

**Comment by varenick** [comment]
@soumith As an alternative, it is possible to implement a helper function `infer_output_shape` which takes an instance of `nn.Module` and input shape, constructs a dummy-tensor with a desired shape (with batch dimension set to 1), calls `forward` on this dummy-tensor, and returns an output shape.

**Comment by ipod825** [comment]
> Implementing infer_output_shape for every class inherited from nn.Module is an unacceptable burden for Module writers. It's easy for nn.Linear, but not so easy for others.

Isn't this exactly the reason why we would like to have the shape inference mechanism?
For example, if only static graph is concerned and every module can infer shape from the input, then [convolution layers with `same` padding](https://github.com/pytorch/pytorch/issues/3867) #3867 can be done automatically without user calculating himself/herself.

I've create a Conv2dSame layer in an ad-hoc way like what @varenick described (use an additional forward pass to inference the shape):

```python
class Conv2dSame(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size,
                 stride=1,
                 padding=0,
                 dilation=1,
                 groups=1,
                 bias=True):
        super(Conv2dSame, self).__init__()
        args = locals()
        args.pop('self')
        args.pop('__class__')
        self.args = args
        self.net = None

    def forward(self, x):
        if self.net is None:
            self.args['padding'] = calc_pad(
                x.shape[1], self.args['kernel_size'], self.args['stride'])[0]
            self.net = nn.Conv2d(**self.args).to(x.device.type)
        return self.net(x)
```

**Comment by albanD** [comment]
This has now been added in https://github.com/pytorch/pytorch/pull/44538
